### ISSUE TITLE
The ZJU MoCap dataset (not refine) has poor training effect.

### ISSUE THREAD

**Issue by 120829** [Yifever20002/R3Avatars#1]
I trained with the ZJU MoCap dataset (not refine) but it works poorly, what's going on?
![Image](https://github.com/user-attachments/assets/427f95ec-92a2-495a-bed8-ec4be1b5f6bb)
The training effect of ZJU MoCap dataset (refine) is normal.
![Image](https://github.com/user-attachments/assets/75efdfaa-be8b-447f-9fcb-deeb655f8f79)

**Comment by Yifever20002** [comment]
Thank you for your interest! Could you provide more details on the training, like how many frames, cameras, and iterations you've trained? We experimentally found that the ZJU_MoCap without refinement will have poor camera calibration, but not like this in your picture (which is blurred).

**Comment by Yifever20002** [comment]
Notice that we mainly tested our code for the refined version in our paper. For the original one you show, the human pose seems to be wrong. Looking forward to more details from you!

**Comment by 120829** [comment]
> Thank you for your interest! Could you provide more details on the training, like how many frames, cameras, and iterations you've trained? We experimentally found that the ZJU_MoCap without refinement will have poor camera calibration, but not like this in your picture (which is blurred).感谢您的关注！您能否提供更多关于训练的细节，比如训练了多少帧、多少个摄像头和迭代次数？我们实验发现，未经优化的 ZJU_MoCap 将会有较差的相机标定，但不像您图片中的那样（模糊）。

I use all the training configuration the same as the refine dataset configuration in the codebase, only modifying the smpl_vertices and smpl_params in the dataset_readers.py to new_vertices, new_params, and the code for mask reading (because the refine and not refine mask formats are different). The same modification works well in GauHuman's code.

**Comment by Yifever20002** [comment]
Could you paste the **def readCamerasZJUMoCapRefine** here (after modified)? Maybe I can help you for debugging it.

**Comment by 120829** [comment]
> Could you paste the **def readCamerasZJUMoCapRefine** here (after modified)? Maybe I can help you for debugging it.

Sorry, I found a bug and missed some code, which has been fixed and currently renders about the same as the refine dataset. Just in case, here's readCamerasZJUMoCapRefine:
```python
def readCamerasZJUMoCapRefine(path, output_view, white_background, image_scaling=0.5, split='train', args=None, mono_test=False, render_novel_pose=False, sim_type=None, pose_step=-1, novel_view_vis=False):
    cam_infos = []

    if mono_test and render_novel_pose:
        if split == 'train':
            pose_start = 0
            pose_interval = 300
            pose_num = 1
        elif split == 'test':
            pose_start = 300
            pose_interval = 1
            pose_num = 200
    elif mono_test:
        if split == 'train':
            pose_start = 0
            pose_interval = 300
            pose_num = 1
        elif split == 'test':
            pose_start = 0
            pose_interval = 1
            pose_num = 300
    elif render_novel_pose:
        if split == 'train':
            pose_start = 0
            pose_interval = 300
            pose_num = 1
        elif split == 'test':
            pose_start = 300
            pose_interval = 20
            pose_num = 10
    else:
        if split == 'train':
            pose_start = 0
            pose_interval = 5
            pose_num = 60
        elif split == 'test':
            pose_start = 0
            pose_interval = 50
            pose_num = 6


    ann_file = os.path.join(path, 'annots.npy')
    annots = np.load(ann_file, allow_pickle=True).item()
    cams = annots['cams']
    ims = np.array([
        np.array(ims_data['ims'])[output_view]
        for ims_data in annots['ims']
    ])

    cam_inds = np.array([
        np.arange(len(ims_data['ims']))[output_view]
        for ims_data in annots['ims']
    ])

    smpl_model = SMPL(sex='neutral', model_dir='./assets/SMPL_NEUTRAL.pkl')

    # SMPL in canonical space
    big_pose_smpl_param = {}
    big_pose_smpl_param['R'] = np.eye(3).astype(np.float32)
    big_pose_smpl_param['Th'] = np.zeros((1,3)).astype(np.float32)
    big_pose_smpl_param['shapes'] = np.zeros((1,10)).astype(np.float32)
    big_pose_smpl_param['poses'] = np.zeros((1,72)).astype(np.float32)
    big_pose_smpl_param['poses'][0, 5] = 45/180*np.array(np.pi)
    big_pose_smpl_param['poses'][0, 8] = -45/180*np.array(np.pi)
    big_pose_smpl_param['poses'][0, 23] = -30/180*np.array(np.pi)
    big_pose_smpl_param['poses'][0, 26] = 30/180*np.array(np.pi)

    big_pose_xyz, _ = smpl_model(big_pose_smpl_param['poses'], big_pose_smpl_param['shapes'].reshape(-1))
    big_pose_xyz = (np.matmul(big_pose_xyz, big_pose_smpl_param['R'].transpose()) + big_pose_smpl_param['Th']).astype(np.float32)

    # obtain the original bounds for point sampling
    big_pose_min_xyz = np.min(big_pose_xyz, axis=0)
    big_pose_max_xyz = np.max(big_pose_xyz, axis=0)
    big_pose_min_xyz -= 0.05
    big_pose_max_xyz += 0.05
    big_pose_world_bound = np.stack([big_pose_min_xyz, big_pose_max_xyz], axis=0)

    if render_novel_pose and split != 'train':
        assert pose_step != -1, "please set pose step!"
        pose_embedding = []
        for pose_index in range(0, 300, 5): # training pose

            i_n = int(os.path.basename(ims[pose_index][0])[:-4])
            smpl_param_path = os.path.join(path, "new_params", '{}.npy'.format(i_n))
            smpl_param = np.load(smpl_param_path, allow_pickle=True).item()
            smpl_param['poses'] = np.array(smpl_param['poses']).astype(np.float32)

            cur_pose_embedding = torch.tensor(smpl_param['poses']).reshape(1, -1, 3)
            cur_mat = axis_angle_to_matrix(cur_pose_embedding.squeeze())
            for i in range(pose_step):
                lst_model_idx = max(0, pose_index-5*(i+1))
                i_n_lst = int(os.path.basename(ims[lst_model_idx][0])[:-4])
                lst_smpl_param_path = os.path.join(path, "new_params", '{}.npy'.format(i_n_lst))
                lst_smpl_param = np.load(lst_smpl_param_path, allow_pickle=True).item()
                lst_smpl_param['poses'] = np.array(lst_smpl_param['poses']).astype(np.float32)
                lst_pose_embedding = torch.tensor(lst_smpl_param['poses']).reshape(1, -1, 3).squeeze()
                lst_mat = axis_angle_to_matrix(lst_pose_embedding) #(B,3,3)
                # calculate delta pose
                cur_delta_mat = torch.bmm(cur_mat, torch.linalg.inv(lst_mat)) #(B,3,3)
                cur_delta_pose = matrix_to_axis_angle(cur_delta_mat)
                # update embedding
                cur_pose_embedding = torch.cat([cur_pose_embedding, cur_delta_pose.unsqueeze(0)], dim=0)
                cur_mat = lst_mat
            cur_pose_embedding = cur_pose_embedding.transpose(0, 1)
            cur_pose_embedding = cur_pose_embedding.reshape(cur_pose_embedding.shape[0], -1)
            pose_embedding.append(cur_pose_embedding.unsqueeze(0))
        pose_embedding = torch.cat(pose_embedding, dim=0)  # [100, 25, 3*(1+step)]
        
        pose_body_parts = {}
        for name in body_parts_smpl:
            pose_part = pose_embedding[:, body_parts_smpl[name], :]
            pose_body_parts[name] = pose_part


    idx = 0
    pair_t_lst = []
    for pose_index in tqdm(range(pose_start, pose_start+pose_num*pose_interval, pose_interval)):
        
        i = int(os.path.basename(ims[pose_index][0])[:-4])
        vertices_path = os.path.join(path, 'new_vertices', '{}.npy'.format(i))
        xyz = np.load(vertices_path).astype(np.float32)

        smpl_param_path = os.path.join(path, "new_params", '{}.npy'.format(i))
        smpl_param = np.load(smpl_param_path, allow_pickle=True).item()
        Rh = smpl_param['Rh']
        smpl_param['R'] = cv2.Rodrigues(Rh)[0].astype(np.float32)
        smpl_param['Th'] = smpl_param['Th'].astype(np.float32)
        smpl_param['shapes'] = smpl_param['shapes'].astype(np.float32)
        smpl_param['poses'] = smpl_param['poses'].astype(np.float32)

        pair_t = []
        jump_note = []

        if render_novel_pose and split != 'train':
            jump_note = [-1 for i in range(len(pose_body_parts))]
            cur_pose_embedding = torch.tensor(smpl_param['poses']).reshape(1, -1, 3)   # no global no hand
            cur_mat = axis_angle_to_matrix(cur_pose_embedding.squeeze())
            for i in range(pose_step):
                lst_model_idx = max(0, pose_index-5*(i+1))
                i_n_lst = int(os.path.basename(ims[lst_model_idx][0])[:-4])
                lst_smpl_param_path = os.path.join(path, "new_params", '{}.npy'.format(i_n_lst))
                lst_smpl_param = np.load(lst_smpl_param_path, allow_pickle=True).item()
                lst_smpl_param['poses'] = np.array(lst_smpl_param['poses']).astype(np.float32)
                lst_pose_embedding = torch.tensor(lst_smpl_param['poses']).reshape(1, -1, 3).squeeze()
                lst_mat = axis_angle_to_matrix(lst_pose_embedding) #(B,3,3)
                # calculate delta pose
                cur_delta_mat = torch.bmm(cur_mat, torch.linalg.inv(lst_mat)) #(B,3,3)
                cur_delta_pose = matrix_to_axis_angle(cur_delta_mat)
                # update embedding
                cur_pose_embedding = torch.cat([cur_pose_embedding, cur_delta_pose.unsqueeze(0)], dim=0)
                cur_mat = lst_mat
            cur_pose_embedding = cur_pose_embedding.transpose(0, 1)
            cur_pose_embedding = cur_pose_embedding.reshape(cur_pose_embedding.shape[0], -1).unsqueeze(0)

            p_idx = 0
            for name in body_parts_smpl:
                cur_pose_part = cur_pose_embedding[:, body_parts_smpl[name], :]
                if sim_type == 'axis-angle':
                    pose_body_part = pose_body_parts[name]
                    maxsim_t, sim = find_most_similar(cur_pose_part, pose_body_part)
                elif sim_type == 'quater':
                    cur_pose_part = axis_angle_to_quaternion(cur_pose_part)
                    pose_body_part = axis_angle_to_quaternion(pose_body_parts[name])
                    maxsim_t, sim = find_most_similar(cur_pose_part, pose_body_part)
                elif sim_type == 'euclidean':
                    pose_body_part = pose_body_parts[name]
                    maxsim_t, sim = find_most_similar(cur_pose_part, pose_body_part, sim_type='euclidean')
                maxsim, maxt = maxsim_t     

                if len(pair_t_lst) == 0:      
                    selected_value = maxt[0]
                    pair_t.append(maxt[0])
                else:
                    t_lst = pair_t_lst[p_idx]
                    distances = torch.abs(maxt - t_lst)
                    neighbor_num = len(distances[distances<3])
                    print(t_lst)
                    print(maxt[distances<3])
                    print(neighbor_num)

                    if neighbor_num > 1:
                        w_ = maxsim[distances < 3][1]/(maxsim[distances < 3][0]+maxsim[distances < 3][1])
                        selected_value = w_ * maxt[distances < 3][0] + (1 - w_) * maxt[distances < 3][1]
                    elif neighbor_num == 1:
                        selected_value = maxt[distances < 3][0]
                    else:
                        # jitter
                        jump_note[p_idx] = t_lst
                        selected_value = maxt[0]
                        
                    pair_t.append(selected_value)

                p_idx+=1

            pair_t_lst = pair_t


        # obtain the original bounds for point sampling
        min_xyz = np.min(xyz, axis=0)
        max_xyz = np.max(xyz, axis=0)
        min_xyz -= 0.05
        max_xyz += 0.05
        world_bound = np.stack([min_xyz, max_xyz], axis=0)

        for view_index in range(len(output_view)):

            if novel_view_vis:
                view_index_look_at = view_index
                view_index = 0

            # Load image, mask, K, D, R, T
            image_path = os.path.join(path, ims[pose_index][view_index].replace('\\', '/'))
            image_name = ims[pose_index][view_index].split('.')[0]
            image = np.array(imageio.imread(image_path).astype(np.float32)/255.)

            # msk_path = image_path.replace('images', 'mask').replace('jpg', 'png')
            msk_path = image_path.replace('images', 'mask')
            msk = cv2.imread(msk_path, cv2.IMREAD_GRAYSCALE)
            # msk = imageio.imread(msk_path)
            msk = (msk != 0).astype(np.uint8)

            if not novel_view_vis:
                cam_ind = cam_inds[pose_index][view_index]
                K = np.array(cams['K'][cam_ind])
                D = np.array(cams['D'][cam_ind])
                R = np.array(cams['R'][cam_ind])
                T = np.array(cams['T'][cam_ind]) / 1000.

                image = cv2.undistort(image, K, D)
                msk = cv2.undistort(msk, K, D)
            else:
                pose = np.matmul(np.array([[1,0,0,0], [0,-1,0,0], [0,0,-1,0], [0,0,0,1]]), get_camera_extrinsics_zju_mocap_refine(view_index_look_at, val=True))
                R = pose[:3,:3]
                T = pose[:3, 3].reshape(-1, 1)
                cam_ind = cam_inds[pose_index][view_index]
                K = np.array(cams['K'][cam_ind])

            image[msk == 0] = 1 if white_background else 0

            # change from OpenGL/Blender camera axes (Y up, Z back) to COLMAP (Y down, Z forward)
            w2c = np.eye(4)
            w2c[:3,:3] = R
            w2c[:3,3:4] = T

            # get the world-to-camera transform and set R, T
            R = np.transpose(w2c[:3,:3])  # R is stored transposed due to 'glm' in CUDA code
            T = w2c[:3, 3]

            # Reduce the image resolution by ratio, then remove the back ground
            ratio = image_scaling
            if ratio != 1.:
                H, W = int(image.shape[0] * ratio), int(image.shape[1] * ratio)
                image = cv2.resize(image, (W, H), interpolation=cv2.INTER_AREA)
                msk = cv2.resize(msk, (W, H), interpolation=cv2.INTER_NEAREST)
                K[:2] = K[:2] * ratio

            image = Image.fromarray(np.array(image*255.0, dtype=np.byte), "RGB")

            focalX = K[0,0]
            focalY = K[1,1]
            FovX = focal2fov(focalX, image.size[0])
            FovY = focal2fov(focalY, image.size[1])

            # get bounding mask and bcakground mask
            bound_mask = get_bound_2d_mask(world_bound, K, w2c[:3], image.size[1], image.size[0])
            bound_mask = Image.fromarray(np.array(bound_mask*255.0, dtype=np.byte))
            # print(msk)
            ms = np.array(msk*255.0, dtype=np.uint8)
            bkgd_mask = Image.fromarray(ms)

            cam_infos.append(CameraInfo(uid=idx, pose_id=pose_index, R=R, T=T, K=K, FovY=FovY, FovX=FovX, image=image,
                            image_path=image_path, image_name=image_name, bkgd_mask=bkgd_mask, 
                            bound_mask=bound_mask, width=image.size[0], height=image.size[1], 
                            smpl_param=smpl_param, world_vertex=xyz, world_bound=world_bound, 
                            big_pose_smpl_param=big_pose_smpl_param, big_pose_world_vertex=big_pose_xyz, 
                            big_pose_world_bound=big_pose_world_bound, pair_t=pair_t, jump_note=jump_note, dataset_type='zju'))

            idx += 1
            
    return cam_infos, pose_interval*pose_num
```

**Comment by Yifever20002** [comment]
Great! Glad you found the bug. The code looks correct now!
